### PR TITLE
modem_cellular: add U-Blox LARA-R6 and implement change of baudrate

### PR DESCRIPTION
**Pull request by emillindq**
Some modems, including LARA-R6 and Quectel EG-25, have a default baudrate of 115200 despite the modems being multiple MBit/s. This PR adds an optional state in modem_cellular.c which will be entered if the uart-target-speed is set in the DTS, changing the speed and reconfiguring the UART peripheral before continuing on with the init script.

@bjarki-andreasen This is a draft, error checks and such has to be added, but I'd like your input first before I spend more time on this. Please note however that the git diff here on Github is quite confusing...

Thanks!

### CODE REVIEW

**Comment by bjarki-andreasen** [comment]
The target baudrate should not be in the devicetree, it should be a Kconfig. Additionally, the commands to send to set the baudrate are constant, like the APN in the init script. It is quite a bit simpler to add an additonal optional script alongside `const struct modem_chat_script *periodic_chat_script;` in the config, which if not NULL, is run.

**Comment by emillindq** [comment]
> . It is quite a bit simpler to add an additonal optional script alongside const struct modem_chat_script *periodic_chat_script; in the config, which if not NULL, is run

I have thought about this a lot, and it is indeed the easy solution. I felt however that hardware-specific configurations should come from the DTS, like it's done in the esp32_at driver
image

Anyhow, using Kconfig symbols I was thinking there should be an enable Kconfig, which enables the actual baud rate symbols. But then you'd have to ifdef everywhere in modem_cellular.c wherever the enabled symbols are used. Which is [not so nice](https://github.com/obgm/libcoap/blob/develop/src/coap_io.c).

Anyways, take a look at the latest commit

File: drivers/modem/modem_cellular.c [hunk 1]
```diff
@@ -717,8 +726,67 @@ static int modem_cellular_on_await_power_on_state_enter(struct modem_cellular_da
...
+	switch (evt) {
+	case MODEM_CELLULAR_EVENT_BUS_OPENED:
+		modem_chat_attach(&data->chat, data->uart_pipe);
+		modem_chat_run_script_async(&data->chat, config->set_baudrate_chat_script);
+		break;
+
+	case MODEM_CELLULAR_EVENT_SCRIPT_SUCCESS:
+		/* Let modem reconfigure */
+		k_msleep(CONFIG_MODEM_CELLULAR_NEW_BAUDRATE_DELAY);
+		__fallthrough;
```
Comment (bjarki-andreasen) [inline drivers/modem/modem_cellular.c:773]: Delays er performed using modem_cellular_start_timer() and waiting for the `MODEM_CELLULAR_EVENT_TIMEOUT`, sleeping is not allowed in this driver as it blocks the system work queue

File: drivers/modem/modem_cellular.c [hunk 2]
```diff
@@ -1899,6 +1971,70 @@ MODEM_CHAT_SCRIPT_DEFINE(u_blox_sara_r5_periodic_chat_script,
...
+
+MODEM_CHAT_SCRIPT_DEFINE(u_blox_lara_r6_set_baudrate_chat_script,
+			 u_blox_lara_r6_set_baudrate_chat_script_cmds,
+			 abort_matches, modem_cellular_chat_callback_handler, 1);
+
+MODEM_CHAT_SCRIPT_CMDS_DEFINE(u_blox_lara_r6_init_chat_script_cmds,
+			      MODEM_CHAT_SCRIPT_CMD_RESP_NONE("AT", 100),
+			      MODEM_CHAT_SCRIPT_CMD_RESP_NONE("AT", 100),
+			      MODEM_CHAT_SCRIPT_CMD_RESP_NONE("AT", 100),
+			      MODEM_CHAT_SCRIPT_CMD_RESP_NONE("AT", 100),
```
Comment (bjarki-andreasen) [inline drivers/modem/modem_cellular.c:1992]: No longer needed? if the modem supports auto baudrate detection (that's what the 4 "AT" are for), you don't need to manually change the baudrate :) If auto baud is not supported, remove the redundant "AT"s

File: drivers/modem/modem_cellular.c (hunk 1 above)
Comment (emillindq) [reply drivers/modem/modem_cellular.c]: Makes sense

File: drivers/modem/modem_cellular.c (hunk 2 above)
Comment (emillindq) [reply drivers/modem/modem_cellular.c]: True that!

**Comment by emillindq** [comment]
I threw another small commit into this PR, a CMUX frame size of 127 causes FCS errors. Different checksums even if address, control and length is the same. Must be a bug in the modem, because anything just above or just below doesn't cause FCS errors

File: drivers/modem/modem_cellular.c [hunk 3]
```diff
@@ -762,10 +830,6 @@ static void modem_cellular_run_init_script_event_handler(struct modem_cellular_d
 		modem_cellular_enter_state(data, MODEM_CELLULAR_STATE_CONNECT_CMUX);
 		break;
 
-	case MODEM_CELLULAR_EVENT_SUSPEND:
-		modem_cellular_enter_state(data, MODEM_CELLULAR_STATE_IDLE);
-		break;
-
```
Comment (bjarki-andreasen) [inline drivers/modem/modem_cellular.c:768]: is this removed intentionally?

File: drivers/modem/Kconfig.cellular [hunk 4]
```diff
@@ -55,4 +57,16 @@ config MODEM_CELLULAR_USER_PIPE_BUFFER_SIZES
 	int "The size of the buffers used for each user pipe in bytes."
 	default 128
 
+config MODEM_CELLULAR_NEW_BAUDRATE
+	int "New baudrate to configure modem to, if supported"
+	range 0 4000000
```
Comment (rerickson1) [inline drivers/modem/Kconfig.cellular:62]: a range of 0 doesn't make sense does it? I recommend setting the min to something like 9600.

Comment (emillindq) [reply drivers/modem/Kconfig.cellular]: fixed

**Review by bjarki-andreasen — DISMISSED** [review]
I think this looks good :)

**Review by rerickson1 — CHANGES_REQUESTED** [review]

File: drivers/modem/modem_cellular.c [hunk 5]
```diff
@@ -1899,6 +1975,65 @@ MODEM_CHAT_SCRIPT_DEFINE(u_blox_sara_r5_periodic_chat_script,
...
+			      MODEM_CHAT_SCRIPT_CMD_RESP("", ok_match),
+			      MODEM_CHAT_SCRIPT_CMD_RESP("AT+CGMR", cgmr_match),
+			      MODEM_CHAT_SCRIPT_CMD_RESP("", ok_match),
+			      MODEM_CHAT_SCRIPT_CMD_RESP("AT+CIMI", cimi_match),
+			      MODEM_CHAT_SCRIPT_CMD_RESP("", ok_match),
+			      MODEM_CHAT_SCRIPT_CMD_RESP("AT+CMUX=0,0,5,31", ok_match));
+
+MODEM_CHAT_SCRIPT_DEFINE(u_blox_lara_r6_init_chat_script, u_blox_lara_r6_init_chat_script_cmds,
+			 abort_matches, modem_cellular_chat_callback_handler, 10);
+
```
Comment (rerickson1) [inline drivers/modem/modem_cellular.c:2015]: please check formatting here

File: drivers/modem/Kconfig.cellular [hunk 6]
```diff
@@ -55,4 +57,16 @@ config MODEM_CELLULAR_USER_PIPE_BUFFER_SIZES
 	int "The size of the buffers used for each user pipe in bytes."
 	default 128
 
+config MODEM_CELLULAR_NEW_BAUDRATE
+	int "New baudrate to configure modem to, if supported"
+	range 9600 4000000
+	default 3000000 if DT_HAS_U_BLOX_LARA_R6_ENABLED
+	default 115200
```
[the remark this answers is not in the document]
Comment (emillindq) [reply drivers/modem/Kconfig.cellular:64]: I guess it depends on both the support of the modem, HW design and UART driver. My vision was that the default would be something that probably works everywhere, and if the user desires something else they can change this in their board's Kconfig.defconfig

[the remark this answers is not in the document]
Comment (rerickson1) [reply drivers/modem/Kconfig.cellular:64]: Never mind, `MODEM_BACKEND_UART` already depends on the UART driver, so no need for a depends here.

**Review by rerickson1 — DISMISSED** [review]

File: drivers/modem/Kconfig.cellular [hunk 7]
```diff
@@ -55,4 +57,16 @@ config MODEM_CELLULAR_USER_PIPE_BUFFER_SIZES
...
+
+config MODEM_CELLULAR_NEW_BAUDRATE_DELAY
+	int "Time modem takes to change baudrate"
```
Comment (tomi-font) [inline drivers/modem/Kconfig.cellular:67]:
```suggestion
	int "Time modem takes to change baudrate in milliseconds"
```

File: drivers/modem/modem_cellular.c [hunk 8]
```diff
@@ -717,8 +726,67 @@ static int modem_cellular_on_await_power_on_state_enter(struct modem_cellular_da
...
+	case MODEM_CELLULAR_EVENT_SCRIPT_SUCCESS:
+		/* Let modem reconfigure */
+		modem_cellular_start_timer(data, K_MSEC(CONFIG_MODEM_CELLULAR_NEW_BAUDRATE_DELAY));
+		break;
+	case MODEM_CELLULAR_EVENT_SCRIPT_FAILED:
+		/* Some modems save the new speed on first change, meaning the
+		 * modem is already at the new baudrate, meaning no reply. So
+		 * ignore any failures and continue as if baudrate is already set
+		 */
+		__fallthrough;
```
Comment (tomi-font) [inline drivers/modem/modem_cellular.c:778]: Suggest adding a `LOG_DBG()` in this `case`.

File: drivers/modem/modem_cellular.c [hunk 9]
```diff
@@ -717,8 +726,67 @@ static int modem_cellular_on_await_power_on_state_enter(struct modem_cellular_da
...
 	case MODEM_CELLULAR_EVENT_TIMEOUT:
+		uart_config_get(config->uart, &cfg);
+		cfg.baudrate = CONFIG_MODEM_CELLULAR_NEW_BAUDRATE;
+		uart_configure(config->uart, &cfg);
+
+		modem_chat_release(&data->chat);
+		modem_pipe_attach(data->uart_pipe, modem_cellular_bus_pipe_handler, data);
+		modem_pipe_close_async(data->uart_pipe);
```
Comment (tomi-font) [inline drivers/modem/modem_cellular.c:786]: Would it make more sense to do the releasing/closing before changing the UART configuration? (not sure)

File: drivers/modem/modem_cellular.c [hunk 10]
```diff
@@ -1899,6 +1975,65 @@ MODEM_CHAT_SCRIPT_DEFINE(u_blox_sara_r5_periodic_chat_script,
 			 modem_cellular_chat_callback_handler, 4);
 #endif
 
+#if DT_HAS_COMPAT_STATUS_OKAY(u_blox_lara_r6)
+MODEM_CHAT_SCRIPT_CMDS_DEFINE(u_blox_lara_r6_set_baudrate_chat_script_cmds,
+			      MODEM_CHAT_SCRIPT_CMD_RESP("ATE0", ok_match),
+			      MODEM_CHAT_SCRIPT_CMD_RESP("AT+IPR="
+					STRINGIFY(CONFIG_MODEM_CELLULAR_NEW_BAUDRATE), ok_match),);
```
Comment (tomi-font) [inline drivers/modem/modem_cellular.c:1982]: minor
```suggestion
					STRINGIFY(CONFIG_MODEM_CELLULAR_NEW_BAUDRATE), ok_match));
```

**Review by tomi-font — CHANGES_REQUESTED** [review]
Maybe a release notes entry is in order?

File: drivers/modem/modem_cellular.c (hunk 9 above)
Comment (emillindq) [reply drivers/modem/modem_cellular.c:784]: Seems to work, shouldn't really make any practical difference

Comment (tomi-font) [reply drivers/modem/modem_cellular.c:784]: Curious to know @bjarki-andreasen's viewpoint on this?

File: doc/releases/release-notes-4.0.rst [hunk 11]
```diff
@@ -146,6 +146,9 @@ Drivers and Sensors
 
 * Modem
 
+  * Added support for U-Blox LARA-R6 modem.
```
Comment (tomi-font) [inline doc/releases/release-notes-4.0.rst:149]:
```suggestion
  * Added support for the U-Blox LARA-R6 modem.
```

File: drivers/modem/modem_cellular.c [hunk 12]
```diff
@@ -717,8 +726,77 @@ static int modem_cellular_on_await_power_on_state_enter(struct modem_cellular_da
...
+
+		ret = uart_config_get(config->uart, &cfg);
+		if (ret < 0) {
+			LOG_ERR("Failed to get UART configuration");
+			break;
+		}
+		cfg.baudrate = CONFIG_MODEM_CELLULAR_NEW_BAUDRATE;
+		ret = uart_configure(config->uart, &cfg);
+		if (ret < 0) {
+			LOG_ERR("Failed to set new baudrate");
```
Comment (tomi-font) [inline drivers/modem/modem_cellular.c:794]: Those logs could log the error code to be even more helpful.

**Review by tomi-font — DISMISSED** [review]

**Review by tomi-font — APPROVED** [review]

File: drivers/modem/modem_cellular.c (hunk 12 above)
Comment (emillindq) [reply drivers/modem/modem_cellular.c]: Why not 🙂

**Review by bjarki-andreasen — APPROVED** [review]

**Review by rerickson1 — APPROVED** [review]

**Pull request merged**